### PR TITLE
Add Apple attestation 

### PR DESCRIPTION
**Pull request by Thumimku**
### Proposed changes in this pull request
Issue : https://github.com/wso2/product-is/issues/16836

Implementation of the ClientAttestationValidator interface specific to Apple attestation.
This class validates attestation responses from Apple devices, ensuring the integrity and authenticity of the attested information. 
The validation process involves decoding the Base64-encoded attestation object, parsing the CBOR data, and performing various checks on the attestation statement and authentication data. 
Additionally, it validates the certificate chain using the Apple Root CA and ensures that the reply party Id matches the configured Apple App ID.
This method developed using following documentation  [Validating Apps That Connect to Your Servers](https://developer.apple.com/documentation/devicecheck/validating_apps_that_connect_to_your_server)

Two additional config parameters added with this fix.
```
<AppleAttestationRootCertificatePath>
${carbon.home}/repository/resources/identity/apple/attestation/apple_attestation_root_ca.pem
</AppleAttestationRootCertificatePath>
```
In the context of PKIX (Public Key Infrastructure for X.509), revocation refers to the process of declaring a digital certificate as invalid before its natural expiration date. In Development this is not necessary, hence provide a config to configure revocation.
`<AppleAttestationRevocationCheckEnabled>false</AppleAttestationRevocationCheckEnabled>`

### CODE REVIEW

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/internal/ClientAttestationMgtServiceComponent.java [hunk 1]
```diff
@@ -43,28 +55,102 @@
...
+    private boolean isCertificateExpiringSoon(X509Certificate certificate) {
+
+        Date currentDate = new Date();
+        Date expirationDate = certificate.getNotAfter();
+
+        // Calculate the difference in days
+        long differenceInDays = (expirationDate.getTime() - currentDate.getTime()) / (24 * 60 * 60 * 1000);
+
+        // Check if the certificate is expiring within a month
+        return differenceInDays <= 30;
```
Comment (janakamarasena) [inline components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/internal/ClientAttestationMgtServiceComponent.java:146]: let's get these numbers to constants. And can you check what the ideal timeframe would be for a such warn log.

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/services/ClientAttestationServiceImpl.java [hunk 2]
```diff
@@ -111,28 +122,64 @@ public ClientAttestationContext validateAttestation(String attestationObject,
...
+     * Checks if the provided attestation object is an Apple Attestation.
+     * This method developed using following documentation
+     * <a href="https://developer.apple.com/documentation/devicecheck/validating_apps_that_connect_to_your_server">
+     *     Validating Apps That Connect to Your Servers
+     * </a>
+     * @param attestationObject The attestation object to be checked.
+     * @return true if it is an Apple Attestation, false otherwise.
+     */
+    private boolean isAppleAttestation(String attestationObject) {
+        // Create a CBOR factory and an ObjectMapper for CBOR serialization.
```
Comment (janakamarasena) [inline components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/services/ClientAttestationServiceImpl.java:150]: Add new line

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+ */
+public class AppleAttestationValidator implements ClientAttestationValidator {
+
+    private static final Log LOG = LogFactory.getLog(AppleAttestationValidator.class);
+    private static final String IOS = "IOS";
+    private String applicationResourceId;
+    private String tenantDomain;
+
+    private ClientAttestationMetaData clientAttestationMetaData;
+    public AppleAttestationValidator(String applicationResourceId,
```
Comment (janakamarasena) [inline components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java:82]: add a new line

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                    new ByteArrayInputStream(x5c.get(0)));
+            X509Certificate caCert = (X509Certificate) certificateFactory.generateCertificate(
+                    new ByteArrayInputStream(x5c.get(1)));
+
+            // Create a CertPath
+            List<X509Certificate> certs = Arrays.asList(credCert, caCert);
+            CertPath certPath = certificateFactory.generateCertPath(certs);
+
+            // Create a CertPathValidator and validate the certificate chain
+            CertPathValidator certPathValidator = CertPathValidator.getInstance("PKIX");
```
Comment (janakamarasena) [inline components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java:176]: Use a constant for "PKIX"

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        Object authDataObject = cborMap.get(AUTH_DATA);
+
+        if (!(authDataObject instanceof byte[])) {
+            clientAttestationContext.setAttested(false);
+            clientAttestationContext.setValidationFailureMessage("Attestation Auth data validation failed. " +
+                    "Auth data is not in expected format.");
+            return false;
+        }
+        byte[] authData = (byte[]) authDataObject;
+        byte[] rpIdHash = Arrays.copyOfRange(authData, 0, 32);
```
Comment (janakamarasena) [inline components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java:224]: Lets add a comment on what this range is 0, 32

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    private boolean isCertificateExpiringSoon(X509Certificate certificate) {
+
+        Date currentDate = new Date();
+        Date expirationDate = certificate.getNotAfter();
+
+        // Calculate the difference in days
+        long differenceInDays = (expirationDate.getTime() - currentDate.getTime()) / (24 * 60 * 60 * 1000);
+
+        // Check if the certificate is expiring within month.
+        return differenceInDays <= 30;
```
Comment (janakamarasena) [inline components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java:283]: Use the same constants that will be defined for https://github.com/wso2/carbon-identity-framework/pull/5165/files#r1393887838

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/internal/ClientAttestationMgtServiceComponent.java (hunk 1 above)
Comment (Thumimku) [reply components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/internal/ClientAttestationMgtServiceComponent.java]: checked with team, 90 days is the best practise

File: components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java (hunk 6 above)
Comment (Thumimku) [reply components/client-attestation-mgt/org.wso2.carbon.identity.client.attestation.mgt/src/main/java/org/wso2/carbon/identity/client/attestation/mgt/validators/AppleAttestationValidator.java]: checked with team, 90 days is the best practise

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**